internal/server/rewrapping.go

```diff
@@ -0,0 +1,92 @@
+package server
+
+import (
+	"context"
+
+	"github.com/hashicorp/boundary/internal/db"
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/boundary/internal/kms"
+)
+
+func init() {
+	kms.RegisterTableRewrapFn("worker_auth_ca_certificate", workerAuthCertRewrapFn)
+	kms.RegisterTableRewrapFn("worker_auth_authorized", workerAuthRewrapFn)
+}
+
+func workerAuthCertRewrapFn(ctx context.Context, dataKeyVersionId string, scopeId string, reader db.Reader, writer db.Writer, kmsRepo *kms.Kms) error {
+	const op = "server.workerAuthCertRewrapFn"
+	var certs []*RootCertificate
+	// indexes on public id, state. neither of which are queryable via scope. this is the fastest query
+	if err := reader.SearchWhere(ctx, &certs, "key_id=?", []interface{}{dataKeyVersionId}, db.WithLimit(-1)); err != nil {
+		return errors.Wrap(ctx, err, op)
+	}
+	wrapper, err := kmsRepo.GetWrapper(ctx, scopeId, kms.KeyPurposeDatabase)
+	if err != nil {
+		return errors.Wrap(ctx, err, op)
+	}
+	newKeyVersionId, err := wrapper.KeyId(ctx)
+	if err != nil {
+		return errors.Wrap(ctx, err, op)
+	}
+	for _, cert := range certs {
+		privateKey, err := decrypt(ctx, cert.PrivateKey, wrapper)
+		if err != nil {
+			return errors.Wrap(ctx, err, op)
+		}
+		if cert.PrivateKey, err = encrypt(ctx, privateKey, wrapper); err != nil {
+			return errors.Wrap(ctx, err, op)
+		}
+		cert.KeyId = newKeyVersionId
+		if _, err := writer.Update(ctx, cert, []string{"PrivateKey", "KeyId"}, nil); err != nil {
+			return errors.Wrap(ctx, err, op)
+		}
+	}
+	return nil
+}
+
+func workerAuthRewrapFn(ctx context.Context, dataKeyVersionId, scopeId string, reader db.Reader, writer db.Writer, kmsRepo *kms.Kms) error {
+	const op = "server.workerAuthRewrapFn"
+	var auths []*WorkerAuth
+	// an index exists on (worker_id, state), so we can query workers via scope and refine with key id. this is the fastest query
+	rows, err := reader.Query(ctx, workerAuthRewrapQuery, []interface{}{dataKeyVersionId, scopeId})
+	if err != nil {
+		return errors.Wrap(ctx, err, op)
+	}
+	for rows.Next() {
+		workerAuth := allocWorkerAuth()
+		if err := rows.Scan(
+			&workerAuth.WorkerKeyIdentifier,
+			&workerAuth.ControllerEncryptionPrivKey,
+			&workerAuth.KeyId,
+		); err != nil {
+			_ = rows.Close()
```

Comment:
This is fine, but the general convention is to just defer this call. I only mention it because it's worth having consistency. You can keep this if you like, but maybe check other places in this package where we iterate rows and see what's done there.

Reply:
wouldn't deferring it wait until the function exits? seems like we would want to close this before we move onto the rewrapping loop

Reply:
It would indeed, but it's only important to call in the error cases, in the cases where the loop exists and rows.Err doesn't error, it's not important to have it called at any particular time.

Reply:
would depend on what the close func does. if we don't call it do we leave the connection open the whole time? might take up (admittedly, a small amount of) db resources deferring it

Reply:
The behavior of the Close function is well documented, which is why I can state this confidently 😛: https://pkg.go.dev/database/sql#Rows.Close.

Reply:
super cool, I'll update it then